### PR TITLE
Add docker support for fast-jar

### PR DESCRIPTION
**Pull request by stuartwdouglas**

### CODE REVIEW

**Comment by geoand** [comment]
Cool, let me check out the PR locally and play with it.

File: core/deployment/src/main/java/io/quarkus/deployment/mutability/ReaugmentTask.java [hunk 1]
```diff
@@ -1,7 +1,6 @@
 package io.quarkus.deployment.mutability;
 
-import static io.quarkus.deployment.pkg.steps.JarResultBuildStep.BUILD_SYSTEM_PROPERTIES;
-import static io.quarkus.deployment.pkg.steps.JarResultBuildStep.DEPLOYMENT_QUARKUS;
+import static io.quarkus.deployment.pkg.steps.JarResultBuildStep.*;
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/mutability/ReaugmentTask.java:3]: We probably don't want this

File: extensions/container-image/container-image-docker/deployment/src/main/java/io/quarkus/container/image/docker/deployment/DockerProcessor.java [hunk 2]
```diff
@@ -210,6 +216,8 @@ private DockerfilePaths getDockerfilePaths(DockerConfig dockerConfig, boolean fo
         } else {
             if (dockerConfig.dockerfileJvmPath.isPresent()) {
                 return ProvidedDockerfile.get(Paths.get(dockerConfig.dockerfileJvmPath.get()), outputDirectory);
+            } else if (packageConfig.type.equals(PackageConfig.FAST_JAR)) {
```
Comment (geoand) [inline extensions/container-image/container-image-docker/deployment/src/main/java/io/quarkus/container/image/docker/deployment/DockerProcessor.java:219]: :+1:

**Comment by geoand** [comment]
Something is not properly built with the docker image when using the `fast-jar`. I'll get back to you soon with more details

File: core/deployment/src/main/java/io/quarkus/deployment/mutability/ReaugmentTask.java (hunk 1 above)
Comment (stuartwdouglas) [reply core/deployment/src/main/java/io/quarkus/deployment/mutability/ReaugmentTask.java:3]: oops, I am using a different computer and the IDE was not fully setup properly.

File: devtools/platform-descriptor-json/src/main/resources/templates/dockerfile-fast-jar.ftl [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    && chown 1001:root /deployments \
+    && curl https://repo1.maven.org/maven2/io/fabric8/run-java-sh/${RUN_JAVA_VERSION}/run-java-sh-${RUN_JAVA_VERSION}-sh.sh -o /deployments/run-java.sh \
+    && chown 1001 /deployments/run-java.sh \
+    && chmod 540 /deployments/run-java.sh \
+    && echo "securerandom.source=file:/dev/urandom" >> /etc/alternatives/jre/lib/security/java.security
+
+# Configure the JAVA_OPTIONS, you can add -XshowSettings:vm to also display the heap size.
+ENV JAVA_OPTIONS="-Dquarkus.http.host=0.0.0.0 -Djava.util.logging.manager=org.jboss.logmanager.LogManager"
+
+COPY ${build_dir}/quarkus-app/lib/* /deployments/lib/
```
Comment (geoand) [inline devtools/platform-descriptor-json/src/main/resources/templates/dockerfile-fast-jar.ftl:48]: This needs to be:

```suggestion
COPY ${build_dir}/quarkus-app/lib/ /deployments/lib/
```

Othewise Docker doesn't preserve the directory structure and running the container doesn't work

**Review by geoand — CHANGES_REQUESTED** [review]

**Comment by geoand** [comment]
Also, the maven tests seem to be failing:

```
2020-06-16T08:55:29.3568304Z [INFO] Running io.quarkus.devtools.commands.AddGradleExtensionsTest
2020-06-16T08:55:29.5582885Z [ERROR] Tests run: 14, Failures: 0, Errors: 14, Skipped: 0, Time elapsed: 0.181 s <<< FAILURE! - in io.quarkus.devtools.commands.AddGradleExtensionsTest
2020-06-16T08:55:29.5583636Z [ERROR] testRegexpMatches  Time elapsed: 0.041 s  <<< ERROR!
2020-06-16T08:55:29.5584340Z java.lang.IllegalStateException: Failed to load resource templates/dockerfile-fast-jar.ftl
```

File: devtools/platform-descriptor-json/src/main/resources/templates/dockerfile-fast-jar.ftl (hunk 3 above)
Comment (stuartwdouglas) [reply devtools/platform-descriptor-json/src/main/resources/templates/dockerfile-fast-jar.ftl:48]: Oops, I fixed this in the local project I was testing with, but forgot to update the template

File: core/deployment/src/main/java/io/quarkus/deployment/mutability/ReaugmentTask.java (hunk 1 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/mutability/ReaugmentTask.java:3]: Seems like your IDE is still messing with you :)

**Review by geoand — APPROVED** [review]

**Comment by geoand** [comment]
Do you want me to fix the CI issue and push into your branch so you can get some rest @stuartwdouglas :)?

**Comment by stuartwdouglas** [comment]
If you want. I had a quick look and I think it is likely to be problem in the tests infra rather than actually being a dev mode problem. I think some of the generated protobuf stuff does not end up in the correct ClassLoader.

**Comment by geoand** [comment]
This is a different error, the one you mentioned were on another PR :)

Here is the issue is:

```
2020-06-16T12:36:21.4792767Z [INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.048 s - in io.quarkus.devtools.commands.RemoveMavenExtensionsTest
2020-06-16T12:36:21.4797884Z [INFO] Running io.quarkus.devtools.project.codegen.rest.BasicRestProjectGeneratorTest
2020-06-16T12:36:22.2413832Z [ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.759 s <<< FAILURE! - in io.quarkus.devtools.project.codegen.rest.BasicRestProjectGeneratorTest
2020-06-16T12:36:22.2414722Z [ERROR] generateFilesWithJaxRsResource  Time elapsed: 0.63 s  <<< FAILURE!
2020-06-16T12:36:22.2415157Z org.mockito.exceptions.verification.TooManyActualInvocations: 
2020-06-16T12:36:22.2415480Z 
2020-06-16T12:36:22.2415877Z projectWriter.write(
2020-06-16T12:36:22.2417213Z     <any string>,
2020-06-16T12:36:22.2417382Z     <any string>
2020-06-16T12:36:22.2417683Z );
2020-06-16T12:36:22.2417947Z Wanted 11 times:
2020-06-16T12:36:22.2418765Z -> at io.quarkus.devtools.project.codegen.rest.BasicRestProjectGeneratorTest.generateFilesWithJaxRsResource(BasicRestProjectGeneratorTest.java:101)
```

**Comment by stuartwdouglas** [comment]
I have updated the test

**Pull request merged**